### ISSUE TITLE
LMMs-Engine Upgrade Plan

### ISSUE THREAD

**Issue by kcz358** [EvolvingLMMs-Lab/lmms-engine#135]
# LMMs-Engine v0.1.4 Upgrade Plan and further feature enhancement todos

### Overview

This issue outlines the upgrade roadmap for the next major version of LMMs-Engine, focusing on Transformers 5.0 upgrade, distributed training capabilities, MOE model support enhancements, and critical bug fixes.

---

#### 1. Transformers 5.0 Upgrade with MOE Training Unification

**Description:** Upgrade transformers dependency to version 5.0 to align with upstream changes and unify MOE training implementation.

**Scope:**
- Update `transformers==4.57.1` to `transformers==5.0.0` (or latest 5.x) in `pyproject.toml`
- Migrate all MOE models to use **stacked parameters** approach
- Update MoE expert routing and parameter management to follow transformers 5.0 patterns
- **Affected Models:** All MOE models except `qwen3_vl_moe`:
  - `qwen3_moe`
  - `qwen3_omni_moe`
  - Other MOE variants in the codebase

**Expected Changes:**
- Review transformers 5.0 release notes and breaking changes
- Update model implementations to use stacked parameters
- Ensure backward compatibility for existing checkpoints where possible
- Update all MOE-related monkey patches and optimizations
- Test with existing MOE training scripts (e.g., `test/train/qwen3_moe/`, `test/train/qwen3_omni_moe/`)

**Files to Modify:**
- `pyproject.toml:19` - Update transformers version
- `src/lmms_engine/models/qwen3_moe/*`
- `src/lmms_engine/models/qwen3_omni_moe/*`
- All other MOE model directories

---

#### 2. Distributed Checkpoint (DCP) for Model Parameter Storage

**Description:** Implement PyTorch Distributed Checkpoint (DCP) for model parameter storage, enabling seamless sharding and continuation across different GPU configurations.

**Scope:**
- Replace current checkpoint saving with `torch.distributed.checkpoint` (DCP)
- Support cross-shard checkpoint saving/loading
- Enable checkpoint resumption with different GPU numbers
- Maintain compatibility with existing FSDP2 merger for consolidated checkpoints

**Expected Changes:**
- Integrate DCP into `fsdp2_utils.py:83` (currently has partial imports)
- Update checkpoint save/load logic in trainers
- Ensure DCP properly handles model sharding (TP, EP, SP, DP)
- Add configuration options for DCP vs traditional checkpointing
- Update documentation and examples

**Files to Modify:**
- `src/lmms_engine/utils/fsdp2_utils.py`
- `src/lmms_engine/train/runner.py` (or trainer implementations)
- `src/lmms_engine/merger/fsdp2.py` - Ensure compatibility with DCP checkpoints

**Benefits:**
- Flexible checkpoint continuation across different GPU counts
- Better support for large-scale distributed training
- Faster checkpoint loading with partial loading capabilities

---

#### 3. MOE Model Support in Merger Class

**Description:** Extend the `CheckpointMerger` infrastructure to properly support MOE model checkpoint merging.

**Scope:**
- Create MOE-specific merger implementation (or enhance existing FSDP2Merger)
- Handle expert parameter sharding during merge
- Support both regular and EMA checkpoints for MOE models
- Ensure compatibility with DCP checkpoints

**Expected Changes:**
- Add MOE-aware merging logic to handle expert parameters
- Update `merger/base.py` interface if needed
- Create specialized merger for MOE models (e.g., `MoEMerger`)
- Add tests for MOE checkpoint merging

**Files to Modify:**
- `src/lmms_engine/merger/base.py`
- `src/lmms_engine/merger/fsdp2.py` - Enhance or add MOE-specific handling
- `src/lmms_engine/merger/__init__.py` - Export new merger

---

#### Fix #133: Loss Normalization in Bagel Training

**Issue:** The loss normalization uses local token counts instead of global token counts across all GPUs. The `dist.all_reduce` calls are commented out in `bagel.py:380`.

**Solution:**
- Uncomment and properly implement global token normalization
- Ensure all_reduce operations are correctly synchronized
- Test with multi-GPU training scenarios

**File:** `src/lmms_engine/models/bagel/bagel.py:380`

---

#### Fix #132: Qwen3-VL Video Timestamp Token Position

**Issue:** Token position mismatch between transformers processor and lmms-engine. The processor outputs `<2.0 seconds><|vision_start|><|vision_pad|>...` but lmms-engine decodes as `<|vision_start|><2.0 seconds><|vision_pad|>...`. Also, image patch size (16) is not specified correctly in the dataset processor.

**Solution:**
- Align timestamp token positioning with transformers upstream implementation
- Fix image patch size configuration in dataset processor
- Ensure video data processing matches transformers expectations

**Files:**
- `src/lmms_engine/datasets/processor/qwen3_vl_processor.py`
- Video timestamp handling in qwen3_vl models

---

#### Fix #128: Qwen Training Issues

**Issue:**
- `model.visual.blocks` should be `vision_model.blocks`
- `processor.image_processor.max_pixels` is not used in some transformers versions

**Solution:**
- Update attribute references to use correct naming
- Ensure processor configuration is properly applied
- Add compatibility layer for different transformers versions

**Files:**
- `src/lmms_engine/models/qwen2_5_vl/monkey_patch.py:134-139`
- `src/lmms_engine/datasets/processor/qwen3_vl_processor.py:19-33`

---

#### 1. Tensor Parallelism Support for Qwen3-VL

**Description:** Implement TP (Tensor Parallelism) for Qwen3-VL and Qwen3-VL-MoE models.

**Scope:**
- Add TP support for `qwen3_vl` models
- Implement combined EP+TP for `qwen3_vl_moe` models
- Update process group manager to handle TP+EP mesh
- Test with various TP and EP configurations

**Reference:** `src/lmms_engine/utils/muon_utils.py:280,298` - Already has some TP+EP mesh preparation for Muon

---

#### 2. Async Tensor Parallelism

**Description:** Enable asynchronous operations during tensor parallelism to improve performance.

**Scope:**
- Implement async TP for all relevant models
- Use async operations for communication overlap
- Optimize TP communication patterns

**Reference:** `src/lmms_engine/parallel/sequence_parallel/ulysses.py:228` - Has `async_op` in context

---

#### 3. Decouple Liger Kernel and RMPad as Independent Monkey Patches

**Description:** Currently, liger kernel optimizations require RMPad (`use_rmpad` must be enabled together). This coupling should be removed by making RMPad a standalone monkey patch type that can be independently applied.

**Current State:**
- Most models check `if use_rmpad:` conditions within liger kernel patching functions
- RMPad is tightly integrated with liger kernel implementations

**Goal:**
- Make RMPad a separate monkey patch type registered with `MONKEY_PATCHER`
- Allow independent application of RMPad and liger kernel patches
- Enable flexible combination: apply liger kernel alone, RMPad alone, or both together

**Implementation Approach:**

1. **Create RMPad Monkey Patch Function:**
   - Define separate `apply_rmpad_to_<model>()` functions for each model type
   - Register these functions with `MONKEY_PATCHER` as "rmpad" type
   - Move RMPad-specific logic (attn_forward, decoder_layer_forward, etc.) out of liger kernel functions

2. **Update Monkey Patch Registration:**
   ```python
   # Example for qwen3_vl
   @MONKEY_PATCHER.register("qwen3_vl", "rmpad")
   def apply_rmpad_to_qwen3_vl(model: PreTrainedModel = None) -> None:
       # Apply RMPad optimizations only
       ...

   @MONKEY_PATCHER.register("qwen3_vl", "liger")
   def apply_liger_kernel_to_qwen3_vl(model: PreTrainedModel = None) -> None:
       # Apply liger kernel optimizations only
       ...
   ```

3. **Independent Application:**
   ```python
   # Apply only RMPad
   MONKEY_PATCHER.apply("qwen3_vl", "rmpad", model=model)

   # Apply only liger kernel
   MONKEY_PATCHER.apply("qwen3_vl", "liger", model=model)

   # Apply both
   MONKEY_PATCHER.apply("qwen3_vl", "rmpad", model=model)
   MONKEY_PATCHER.apply("qwen3_vl", "liger", model=model)
   ```

**Files to Modify:**
- All model directories with liger implementations:
  - `src/lmms_engine/models/qwen3_vl/monkey_patch.py`
  - `src/lmms_engine/models/qwen3_moe/monkey_patch.py`
  - `src/lmms_engine/models/qwen3_omni_moe/monkey_patch.py`
  - `src/lmms_engine/models/qwen3/qwen3_ops.py`
  - `src/lmms_engine/models/qwen2_5_vl/monkey_patch.py`
  - And other qwen model variants

**Benefits:**
- Clear separation of concerns
- Flexible optimization combinations
- Easier to add new optimization types in the future
- Better testability for each optimization independently

---

### 📋 Implementation Priority

**Phase 1 (High Priority - Core Functionality):**
1. Fix #133, #132, #128 (Critical bug fixes)
2. Transformers 5.0 upgrade
3. MOE training unification

**Phase 2 (Medium Priority - Infrastructure):**
4. DCP implementation
5. MOE merger support
6. Decouple liger kernel and RMPad

**Phase 3 (Lower Priority - Enhancements):**
7. TP support for Qwen3-VL
8. Async TP

---

### 🧪 Testing Requirements

- **Unit tests** for new DCP functionality
- **Integration tests** for MOE checkpoint merging
- **Multi-GPU tests** for TP and EP configurations
- **Regression tests** for bug fixes
- **Performance benchmarks** for async TP
- **Independent tests** for RMPad and liger kernel patches

---

### 📚 Documentation Updates

- Update migration guide for transformers 5.0
- Document DCP usage and configuration
- Add MOE merger examples
- Update TP/EP configuration documentation
- Document independent usage of RMPad and liger kernel
